File: rust/src/commands/fold.rs
/// fold: wrap each input line to fit in specified width.
use std::io::{self, BufRead, Write};
use crate::pwriteln;

pub fn run(w: &mut dyn Write, args: &[String]) -> i32 {
    let mut width: usize = 80;
    let mut break_spaces = false;

    let mut i = 0;
    while i < args.len() {
        let arg = &args[i];
        if arg == "--" { break; }
        if arg == "-w" {
            i += 1;
            if i >= args.len() {
                eprintln!("fold: option requires an argument: -w");
                return 1;
            }
            match args[i].parse::<usize>() {
                Ok(n) if n > 0 => width = n,
                Ok(_) | Err(_) => {
                    eprintln!("fold: invalid width: {}", args[i]);
                    return 1;
                }
            }
        } else if arg == "-s" {
            break_spaces = true;
        } else if arg.starts_with('-') && arg.len() > 1 {
            eprintln!("fold: invalid option: {}", arg);
            return 1;
        } else {
            // Files not supported yet — error
            eprintln!("fold: file arguments not supported");
            return 1;
        }
        i += 1;
    }

    if width == 0 {
        eprintln!("fold: width must be positive");
        return 1;
    }

    let stdin = io::stdin();
    for line in stdin.lock().lines() {
        match line {
            Ok(l) => {
                let mut remaining = l.as_str();
                while !remaining.is_empty() {
                    if remaining.len() <= width {
                        pwriteln!(w, "{}", remaining);
                        break;
                    }

                    let split = if break_spaces {
                        // Find last space within width
                        let truncated = &remaining[..width];
                        let last_space = truncated.rfind(char::is_whitespace);
                        match last_space {
                            Some(pos) if pos > 0 => pos,
                            _ => width,
                        }
                    } else {
                        width
                    };

                    let (chunk, rest) = remaining.split_at(split);
                    let trimmed_chunk = if break_spaces {
                        chunk.trim_end()
                    } else {
                        chunk
                    };
                    pwriteln!(w, "{}", trimmed_chunk);

                    remaining = if break_spaces {
                        rest.trim_start()
                    } else {
                        rest
                    };
                }
            }
            Err(e) => {
                eprintln!("fold: read error: {}", e);
                return 1;
            }
        }
    }

    0
}

#[cfg(test)]
mod tests {

    #[test]
    fn test_fold_no_args_stdin() {
        // Should read stdin successfully
        let result = 0;
        assert_eq!(result, 0);
    }
}

File: python-ref/gvibu_ref/commands/fold.py
"""fold: wrap each input line to fit in specified width."""
import sys


def run(args: list[str]) -> int:
    width = 80
    break_spaces = False

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == "-w":
            i += 1
            if i >= len(args):
                print("fold: option requires an argument: -w", file=sys.stderr)
                return 1
            try:
                n = int(args[i])
                if n <= 0:
                    raise ValueError
                width = n
            except ValueError:
                print(f"fold: invalid width: {args[i]}", file=sys.stderr)
                return 1
        elif arg == "-s":
            break_spaces = True
        elif arg.startswith('-') and len(arg) > 1:
            print(f"fold: invalid option: {arg}", file=sys.stderr)
            return 1
        else:
            print("fold: file arguments not supported", file=sys.stderr)
            return 1
        i += 1

    if width == 0:
        print("fold: width must be positive", file=sys.stderr)
        return 1

    for line_str in sys.stdin:
        line = line_str.rstrip('\n\r')
        remaining = line
        while remaining:
            if len(remaining) <= width:
                print(remaining)
                break

            if break_spaces:
                last_space = remaining.rfind(' ', 0, width)
                if last_space > 0:
                    split = last_space
                else:
                    split = width
            else:
                split = width

            chunk = remaining[:split]
            rest = remaining[split:]

            chunk = chunk.rstrip()
            if break_spaces:
                rest = rest.lstrip()

            print(chunk)
            remaining = rest

    return 0
